### PR TITLE
Add the MemberActivationMailEvent

### PR DESCRIPTION
**Pull request by fritzmg**
Within `ModuleRegistration` of the `contao/core-bundle` we still have implementations specific to `contao/newsletter-bundle`. This is because we want to provide an additional `##channels##` simple token for the activation email text when newsletter subscriptions are enabled in the registration module. And currently there is no way to manipulate the email message or the provided simple tokens for the email message.

This PR introduces a `MemberActivationMailEvent` with which you can add additional simple tokens for the activation email. You can also change the subject of the email or the whole text - or even prevent sending the activation mail (through `OptInToken::send`) altogether. 

The latter would be useful for `terminal42/notification_center` for example which allows you to send the activation link separately - and currently the Notification Center disables `reg_enable` of the module as a work around, so that the activation email is not sent. With this new event the Notification Center can extract the already provided tokens from the event and provide them in their own notification while also preventing the default email from being sent.

### CODE REVIEW

**Review by aschempp — COMMENTED** [review]
shouldn't we rather not send the email if the event (propagation) is stopped (instead of sendOptInToken)?

**Comment by fritzmg** [comment]
> shouldn't we rather not send the email if the event (propagation) is stopped (instead of sendOptInToken)?

Hmm, you might still want to be able to collect all simple tokens from all registered event listeners. Imagine if the Notification Center stops propagation - but is executed before the Newsletter's listener. Then the `channels` simple token would be missing. Of course the Notification Center will set a very low priority on its own listener so it is more likely to be the last decider. However I still feel like the decision to send or not should be independent from the event's propagation.

**Comment by Toflar** [comment]
Will be 4.13 compatible so this isn't of great use for me just yet 😓

**Comment by fritzmg** [comment]
Yes, but I meant in principle. It's not just for the NC anyway - see also the issue with the newsletter-bundle.

File: core-bundle/src/Event/MemberActivationMailEvent.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        return $this;
+    }
+
+    public function getSimpleTokens(): array
+    {
+        return $this->simpleTokens;
+    }
+
+    public function setSendOptInToken(bool $sendOptInToken): self
```
Comment (Toflar) [inline core-bundle/src/Event/MemberActivationMailEvent.php]: Can we not use setters and getters here but be more speaking?

* disableSendingTheOptInToken(): self
* enableSendingTheOptInToken(): self
* shouldSendOptInToken(): bool

File: core-bundle/src/Event/MemberActivationMailEvent.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public function setSimpleTokens(array $simpleTokens): self
+    {
+        $this->simpleTokens = $simpleTokens;
+
+        return $this;
+    }
+
+    public function setSimpleToken(string $token, string $value): self
```
Comment (Toflar) [inline core-bundle/src/Event/MemberActivationMailEvent.php]: Can we have an `addSimpleToken(string $key, string $value)`?

File: newsletter-bundle/src/EventListener/MemberActivationMailListener.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            return;
+        }
+
+        $channels = $this->connection->fetchFirstColumn('SELECT title FROM tl_newsletter_channel WHERE id IN (?)', [$newsletter], [Types::SIMPLE_ARRAY]);
+
+        if (empty($channels)) {
+            return;
+        }
+
+        $event->setSimpleToken('channels', implode("\n", $channels));
```
Comment (Toflar) [inline newsletter-bundle/src/EventListener/MemberActivationMailListener.php]: This place is where you'd use `addSimpleToken()`.

**Review by Toflar — CHANGES_REQUESTED** [review]
I think the event can be improved :)

File: core-bundle/src/Event/MemberActivationMailEvent.php (hunk 2 above)
Comment (Toflar) [reply core-bundle/src/Event/MemberActivationMailEvent.php]: And a `removeSimpleToken(string $key)`?

**Review by Toflar — DISMISSED** [review]
❤️

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**